src/mono/mono/mini/mini-trampolines.c

```diff
@@ -554,6 +554,15 @@ common_call_trampoline (host_mgreg_t *regs, guint8 *code, MonoMethod *m, MonoVTa
 		} else {
 			MonoObject *this_argument = (MonoObject *)mono_arch_get_this_arg_from_call (regs, code);
 
+#ifdef TARGET_RISCV
+			if (!m_class_is_inited (this_argument->vtable->klass))
```

Comment:
Shouldn't this be done by mono_arch_geta_this_arg_from_call () ?

Reply:
but `mono_arch_geta_this_arg_from_call ()` has been called in three place in `mini-trampolines.c`. are [mini-trampolines.c#L555](https://github.com/Xinlong-Wu/runtime/blob/dd932d886a47aa73110e87a612c337e6b64c58fa/src/mono/mono/mini/mini-trampolines.c#L555) ， [mini-trampolines.c#L810](https://github.com/Xinlong-Wu/runtime/blob/dd932d886a47aa73110e87a612c337e6b64c58fa/src/mono/mono/mini/mini-trampolines.c#L810) and [mini-trampolines.c#L996](https://github.com/Xinlong-Wu/runtime/blob/dd932d886a47aa73110e87a612c337e6b64c58fa/src/mono/mono/mini/mini-trampolines.c#L996) .

In which, only [mini-trampolines.c#L555](https://github.com/Xinlong-Wu/runtime/blob/dd932d886a47aa73110e87a612c337e6b64c58fa/src/mono/mono/mini/mini-trampolines.c#L555) has checked `m_class_is_inited (this_argument->vtable->klass)`. 

I'm not sure should other two place check the `m_class_is_inited (this_argument->vtable->klass)`. So I put it here.

I will put it into `mono_arch_geta_this_arg_from_call ()` if it dosen't matter when we check `m_class_is_inited (this_argument->vtable->klass)` in another place.

Reply:
Why is the m_class_is_inited () check needed then ? If there are objects of a class created, then the class should be inited already.

Reply:
well, it is another story.

RISCV use `a0-a7` reg to store the parameter of function call. but the `a0` will store the address of retuen value when the function returns ArgVtypeByRef. While the `this_pointer` are been set as `a0` forever in `mono_arch_geta_this_arg_from_call`.

Thus, when a function returns `ArgVtypeByRef` and has `this_pointer`, the `a0` will be  address of retuen value and `a1` will be the `this_pointer`. example is here https://godbolt.org/z/vco8ozKvx 

Current `mono_arch_geta_this_arg_from_call` dosen't pass enough information about the return value and this_pointer, it will treat treat `a0` as `this_pointer` forever. That means it will treat address of retuen value as `this_pointer` in previous case. So the current solution is "it might be the address of retuen value if can't get valid class information. Then, get `this_pointer` again"

If current solution is unacceptable, May be I will customize the `mono_arch_geta_this_arg_from_call` for riscv as

```c
gpointer
mono_arch_get_this_arg_from_call (host_mgreg_t *regs, guint8 *code, bool returnVtypeByref)
{
	MonoObject *this = (MonoObject *)regs [returnVtypeByref ? RISCV_A1 : RISCV_A0];
	return (gpointer)this;
}
```

Does the new solution is acceptable?

Reply:
This problem exists on other platforms as well, what we do is we pass the this pointer in the first arg reg, and the vtype return addr in the second one. This is not against the ABI, because these are managed methods who don't have to conform to the ABI calling convention, they will not be called from native code etc. See
https://github.com/dotnet/runtime/blob/main/src/mono/mono/mini/mini-amd64.c#L948
as an example.

Reply:
If current solution is unacceptable, May be I will customize the `mono_arch_geta_this_arg_from_call` for riscv as

```c
gpointer
mono_arch_get_this_arg_from_call (host_mgreg_t *regs, guint8 *code, bool returnVtypeByref)
{
	MonoObject *this = (MonoObject *)regs [returnVtypeByref ? RISCV_A1 : RISCV_A0];
	return (gpointer)this;
}
```
But the problem is, Can I get the Method signature in `common_call_trampoline ()`?

Reply:
> This problem exists on other platforms as well, what we do is we pass the this pointer in the first arg reg, and the vtype return addr in the second one. This is not against the ABI, because these are managed methods who don't have to conform to the ABI calling convention, they will not be called from native code etc. See https://github.com/dotnet/runtime/blob/main/src/mono/mono/mini/mini-amd64.c#L948 as an example.

Great, Will update pr soon )